### ISSUE TITLE
task_for_pid(0) not returning valid port(?)

### ISSUE THREAD

**Issue by OatmealDome** [checkra1n/BugTracker#236]
**What device + iOS version are you on?**

iPad7,5 on iOS 13.2.2, Cydia not installed

**What checkra1n version are you using?**

0.9.1

**What are the steps to reproduce the issue?**
1. Call `task_for_pid(0)`
2. Attempt to do something with the returned `mach_port_t`

**What do you expect, and what is happening instead?**

* Expected: the API call to succeed, for example, ``mach_vm_allocate`` to return memory allocated within the kernel's memory space
* Returned: failed call, for example,`KERN_INVALID_ARGUMENT`

**Any other info, error logs, screenshots, ...?**

Crappy example code:

```
kern_return_t ret;
mach_port_t kernelTask;

ret = task_for_pid(mach_task_self(), 0, &kernelTask);
if (ret != KERN_SUCCESS)
{
    printf("could not get tfp0, ret = 0x%x\n", ret);
    return -1;
}

if (!MACH_PORT_VALID(kernelTask))
{
    printf("tfp0 not valid, ret = 0x%x\n", ret);
    return -1;
}

pid_t pid;
ret = pid_for_task(kernelTask, &pid);
if (ret != KERN_SUCCESS || pid != 0)
{
    printf("pid_for_task returned not kernel task, ret = 0x%x, pid = %d\n", ret, pid);
    return -1;
}

printf("got mach port = 0x%x\n", kernelTask);

mach_vm_address_t address = 0;
ret = mach_vm_allocate(kernelTask, (mach_vm_address_t *)&address, 4096, VM_FLAGS_ANYWHERE);
if (ret != KERN_SUCCESS)
{
    printf("failed to allocate in kernel space, ret = 0x%x\n", ret);
    return -1;
}

printf("OK, address=0x%x", address);
```

Entitlements:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
        <key>get-task-allow</key>
        <true/>
        <key>task_for_pid-allow</key>
        <true/>
        <key>platform-application</key>
        <true/>
        <key>com.apple.private.security.no-container</key>
        <true/>
</dict>
</plist>
```

Output: ``failed to allocate in kernel space, ret = 0x4``

I feel like I'm missing something since I'm still new to kernel hacking, but at the same time I feel it might be a legitimate bug (possibly related to #71?). Any help is appreciated.

**Comment by Halo-Michael** [comment]
Same issues.

**Comment by kpwn** [comment]
this sounds like the task_eval_conversion patch not being fully working. i had tested this on 13.2, but maybe 13.2.2 broke something (13.2 changed the patch a bit, so i wouldn't be surprised)

**Comment by Halo-Michael** [comment]
> this sounds like the task_eval_conversion patch not being fully working. i had tested this on 13.2, but maybe 13.2.2 broke something (13.2 changed the patch a bit, so i wouldn't be surprised)

But I still get the same wrong with checkra1n 0.9.2, on iPhoneSE iOS13.2.

**Comment by Halo-Michael** [comment]
This is the project I am trying to use: https://github.com/0x7ff/dimentio
It use find_task()
![B76E86A8-BBB7-4A59-89F9-F8295ACD433A](https://user-images.githubusercontent.com/45062120/68660716-b0df7080-0574-11ea-9c91-d7114011aba0.png)
to get kernel task, and try to unlock nvram.
But unfortunately, it doesn't work.
![EC91549B-F31B-4BE5-8A60-C03997FE538B](https://user-images.githubusercontent.com/45062120/68660817-db312e00-0574-11ea-902a-bca35c36910d.png)

**Comment by nullpixel** [comment]
task_conversion_eval is seemingly broken on iOS 13.2.2, we’ll have a fix out when we can.

**Comment by Halo-Michael** [comment]
> task_conversion_eval is seemingly broken on iOS 13.2.2, we’ll have a fix out when we can.

https://github.com/checkra1n/BugTracker/issues/236#issuecomment-552817858
It's this because the same reason?

**Comment by OatmealDome** [comment]
fwiw, I just tested the sample code on iOS 13.2 on iPhone8,1 and it failed. Seems that the `task_conversion_eval` patch never worked on iOS 13.2 in the first place?

**Comment by kpwn** [comment]
https://assets.checkra.in/test/checkra1n%20beta%200.9.3.7.dmg should fix this - the patch was indeed no bueno

**Comment by pwmoore** [comment]
Just tried this one on an iPhone 6s+ and 8+, both with 13.2.2. Can confirm that tfp0 works, however, it seems that the 8+ has a problem with this one. It displays the checkra1n boot messages, then displays the Apple logo for a few seconds, before going black and rebooting. I get a panic log with the message "RTBuddy(SMC)::_setManagedStateGated No response received in 20s, SMC not started? (3)". If I revert back to 0.9.3.2, it does boot, albeit with no tfp0.

However, the 6s+ boots fine, and I was able to confirm that tfp0 works there. Haven't tried on any other devices though.

**Comment by kpwn** [comment]
https://assets.checkra.in/test/checkra1n%20beta%200.9.4.dmg please try this, and be sure to follow the DFU instructions in the GUI. Please tell me if this fixes your issues.

**Comment by arx8x** [comment]
> 
> 
> This is the project I am trying to use: [0x7ff/dimentio](https://github.com/0x7ff/dimentio)
> It use find_task()
> ![B76E86A8-BBB7-4A59-89F9-F8295ACD433A](https://user-images.githubusercontent.com/45062120/68660716-b0df7080-0574-11ea-9c91-d7114011aba0.png)
> to get kernel task, and try to unlock nvram.
> But unfortunately, it doesn't work.
> ![EC91549B-F31B-4BE5-8A60-C03997FE538B](https://user-images.githubusercontent.com/45062120/68660817-db312e00-0574-11ea-902a-bca35c36910d.png)

kernel_task isn't the reason it fails. Read the code further. It gets the kernel_task and afaik, only the validation fails. The kernel_task you get is usable. It fails where it tries  to make use of task_info patch which was present in other jailbreaks (not entirely sure about chimera). This patch writes kernel base and kernel slide to task_dyld_info struct of kernel task, allowing you to later call task_info and read these values. Without this, dimentio can't find the boot-nonce string. It doesn't unlock nvram btw. This patch is not implemented in checkra1n yet. You either have to find kbase yourself or wait for the patch to be added. If this worked, nvnonce and system info should work. Nvnonce attempts to iterate through kmem regions to find kbase but it doesn't seem to work on iOS 13. I don't have an iOS 13 device to work around this.

**Comment by arx8x** [comment]
@kpwn 9.4 fixes #324

**Comment by pwmoore** [comment]
Tested 0.9.4 on my 8+, it boots fine and tfp0 is working! Thanks! :)
